File: src/char_gen.c
//*****************************************************************************
//
// char_gen.c
//
// Contains all of the functions neccessary for generating a new character. 
// Different muds will handle this process in wildly different ways. As such,
// I figured it would be good to localize all of the login details to a single 
// file for people to tailor to their specific mud. It also makes updating to 
// newer versions a bit easier. Your character generation process should be
// laid out here. A basic framework is given, but it could use lots of work. In
// fact, it is crap and could use LOTS of work. I encourage people to write 
// their own login procedure and submit their designs to me so I can develop a 
// better general method for later versions of the mud.
//
//*****************************************************************************

#include "mud.h"
#include "utils.h"
#include "socket.h"
#include "account.h"
#include "character.h"
#include "world.h"
#include "save.h"
#include "races.h"
#include "handler.h"
#include "inform.h"



//*****************************************************************************
// mandatory modules
//*****************************************************************************
#include "scripts/script.h"



//*****************************************************************************
// local functions, defines, structures, and the like
//*****************************************************************************

// functions in the char creation procedure
void char_ask_name(SOCKET_DATA *sock, char *arg);
void char_ask_sex (SOCKET_DATA *sock, char *arg);
void char_ask_race(SOCKET_DATA *sock, char *arg);


//
// List all of the races available during creation
void list_races(SOCKET_DATA *sock) {
  send_to_socket(sock, "Available races are:\r\n");
  send_to_socket(sock, "%s", raceGetList(TRUE));
  send_to_socket(sock, "\r\n\r\nPlease enter your choice: ");
}


//
// Character names must be at least 4 characters long, and no more than 12.
// The name must only be composed of alphanumeric characters
bool check_char_name(const char *name) {
  int i, size = strlen(name);
  if (size < 3 || size > 12)
    return FALSE;
  for (i = 0; i < size; i++)
    if (!isalpha(name[i])) 
      return FALSE;
  return TRUE;
}


//
// Query for the character's name.
void char_ask_name(SOCKET_DATA *sock, char *arg) {
  CHAR_DATA *ch = NULL;

  // check for a legal name
  if(!check_char_name(arg))
    text_to_buffer(sock, 
		   "Sorry, that was an illegal name. Please pick another.\r\n"
		   "What is your name? ");
  // a character with this name already exists
  else if(char_exists(arg))
    text_to_buffer(sock,
		   "A character with this name already exists.\r\n"
		   "What is your name? ");
  // we're OK to make this new character
  else {
    ch = newChar();

    // give the player it's name 
    *arg = toupper(*arg);
    charSetName(ch, arg);

    // and room description. In the future, we may want
    // to consider allowing players to set their own rdescs 
    char rdesc[SMALL_BUFFER];
    sprintf(rdesc, "%s is here.", arg);
    charSetRdesc(ch, rdesc);

    // socket <-> player 
    charSetSocket(ch, sock);
    socketSetChar(sock, ch);

    // prepare for next step 
    text_to_buffer(sock, "What is your sex (M/F/N)? ");
    socketReplaceInputHandler(sock, char_ask_sex, NULL);
  }
}


//
// Ask for the character's sex. 
void char_ask_sex (SOCKET_DATA *sock, char *arg) {
  switch(toupper(*arg)) {
  case 'F':
    charSetSex(socketGetChar(sock), SEX_FEMALE);
    break;
  case 'M':
    charSetSex(socketGetChar(sock), SEX_MALE);
    break;
  case 'N':
    charSetSex(socketGetChar(sock), SEX_NEUTRAL);
    break;
  default:
    text_to_buffer(sock, "\r\nInvalid sex. Try again (M/F/N) :");
    return;
  }

  // onto race choosing
  socketReplaceInputHandler(sock, char_ask_race, NULL);
  list_races(sock);
}


//
// Pick a race for our new character
void char_ask_race(SOCKET_DATA *sock, char *arg) {
  if(!arg || !*arg || !raceIsForPC(arg))
    send_to_socket(sock, "Invalid race! Please try again: ");
  else {
    charSetRace(socketGetChar(sock), arg);
    charResetBody(socketGetChar(sock));

    // give the character a unique id 
    int next_char_uid = mudsettingGetInt("puid") + 1;
    mudsettingSetInt("puid", next_char_uid);
    charSetUID(socketGetChar(sock), next_char_uid);
    
    // if it's the first player, set him as the highest level 
    if(charGetUID(socketGetChar(sock)) == 1)
      charSetLevel(socketGetChar(sock), MAX_LEVEL);

    // add the character to the game 
    char_to_game(socketGetChar(sock));
    
    log_string("New player: %s has entered the game.", 
	       charGetName(socketGetChar(sock)));
    
    // and into the game 
    // pop the input handler for char creation and add the one for
    // playing the game
    socketReplaceInputHandler(sock, handle_cmd_input, show_prompt);
    text_to_buffer(sock, motd);
    
    // we should do some checks here to make sure the start room exists
    char_to_room(socketGetChar(sock), worldGetRoom(gameworld, START_ROOM));
    look_at_room(socketGetChar(sock), charGetRoom(socketGetChar(sock)));
    
    // and save him 
    save_player(socketGetChar(sock));
    
    //************************************************************
    //******                 VERY IMPORTANT                 ******
    //************************************************************
    // add this character to our account list
    accountPutChar(socketGetAccount(sock), charGetName(socketGetChar(sock)));
    save_account(socketGetAccount(sock));

    // check enterance scripts
    try_enterance_script(socketGetChar(sock), charGetRoom(socketGetChar(sock)), NULL);
  }
}



//*****************************************************************************
// implementation of char_gen.c
//*****************************************************************************
void start_char_gen(SOCKET_DATA *sock) {
  text_to_buffer(sock, "What is your character's name? ");
  socketPushInputHandler(sock, char_ask_name, NULL);
}

File: lib/pymodules/char_gen.py
'''
char_gen.py

The basic character generation module. Allows accounts to create new characters
with basic selection for name, sex, and race.
'''
import mudsys, mud, socket, char, hooks



def check_char_name(arg):
    '''checks to make sure the character name is valid. Names are valid if they
       are greater than 2 characters, less than 13, and comprise only alpha
       characters.'''
    return len(arg) >= 3 and len(arg) <= 12 and arg.isalpha()

def cg_name_handler(sock, arg):
    if not check_char_name(arg):
        sock.send("{cIllegal name, please pick another.{n\r\n")
    elif mudsys.player_exists(arg):
        sock.send("{cA player with that name already exists.{n\r\n")
    elif mudsys.player_creating(arg):
        sock.send("{cA player is already being created with that name.{n\r\n")
    elif arg.lower().startswith("guest"):
        sock.send("{cCharacter names cannot begin with 'guest'.{n\r\n")
    else:
        name = arg[0].upper() + arg[1:]
        ch = mudsys.create_player(name)

        if ch == None:
            sock.send("{cIllegal name, please pick another.{n\r\n")
        else:
            mudsys.attach_char_socket(ch, sock)
            ch.rdesc = ch.name + " is here."
            sock.pop_ih()

def cg_sex_handler(sock, arg):
    try:
        result = {
            'M' : 'male',
            'F' : 'female',
            }[arg[0].upper()]
        sock.ch.sex = result
        sock.pop_ih()
    except:
        sock.send("{cInvalid sex, try again.{n\r\n")

def cg_race_handler(sock, arg):
    if not mud.is_race(arg, True):
        sock.send("{cInvalid race selection, try again.{n\r\n")
    else:
        sock.ch.race = arg.lower()
        sock.pop_ih()

def cg_finish_handler(sock, arg):
    # pop our input handler for finishing character generation
    sock.pop_ih()

    # log that the character created
    mud.log_string("New player: " + sock.ch.name + " has entered the game.")
    
    # register and save him to disk and to an account
    mudsys.do_register(sock.ch)

    # make him exist in the game for functions to look him up
    mudsys.try_enter_game(sock.ch)

    # run the init_player hook
    hooks.run("init_player", hooks.build_info("ch", (sock.ch,)))
    
    # attach him to his account and save the account
    sock.account.add_char(sock.ch)
    mudsys.do_save(sock.account)
    mudsys.do_save(sock.ch)

    # clear their screen
    sock.ch.act("clear")
    
    # send them the motd
    sock.ch.page(mud.get_motd())

    # make him look at the room
    sock.ch.act("look")

    # run our enter hook
    hooks.run("enter", hooks.build_info("ch rm", (sock.ch, sock.ch.room)))

def cg_name_prompt(sock):
    sock.send_raw("What is your character's name? ")

def cg_sex_prompt(sock):
    sock.send_raw("What is your sex (M/F)? ")

def cg_race_prompt(sock):
    sock.send("Available races are: ")
    sock.send(mud.list_races(True))
    sock.send_raw("\r\nPlease enter your choice: ")

def cg_finish_prompt(sock):
    sock.send_raw("{c*** Press enter to finish character generation:{n ")



################################################################################
# character generation hooks
################################################################################
def char_gen_hook(info):
    '''Put a socket into the character generation menu when character generation
       hooks are called.
    '''
    sock, = hooks.parse_info(info)
    sock.push_ih(mudsys.handle_cmd_input, mudsys.show_prompt, "playing")
    sock.push_ih(cg_finish_handler, cg_finish_prompt)
    sock.push_ih(cg_race_handler, cg_race_prompt)
    sock.push_ih(cg_sex_handler, cg_sex_prompt)
    sock.push_ih(cg_name_handler, cg_name_prompt)

def guest_gen_hook(info):
    sock, = hooks.parse_info(info)
    sock.push_ih(mudsys.handle_cmd_input, mudsys.show_prompt, "playing")

    # make the guest character
    ch = mudsys.create_player("Guest")

    # oops...
    if ch == None:
        sock.send("Sorry, there were issues creating a guest account.")
        sock.close()

    mudsys.attach_char_socket(ch, sock)
    ch.rdesc = "a guest player is here, exploring the world."
    ch.name  = ch.name + str(ch.uid)
    ch.race  = "human"

    # log that the character created
    mud.log_string("Guest character created (id %d)." % ch.uid)

    # make him exist in the game for functions to look him up
    mudsys.try_enter_game(ch)

    # run the init_player hook
    hooks.run("init_player", hooks.build_info("ch", (ch,)))

    # clear our screen
    ch.act("clear")

    # send them the motd
    ch.page(mud.get_motd())
    
    # make him look at the room
    ch.act("look")

    # run our enter hook
    hooks.run("enter", hooks.build_info("ch rm", (ch, ch.room)))



################################################################################
# loading and unloading the module
################################################################################
hooks.add("create_character", char_gen_hook)
hooks.add("create_guest",     guest_gen_hook)

def __unload__():
    '''removes the hooks for character generation'''
    hooks.remove("create_character", char_gen_hook)
    hooks.remove("create_guest",     guest_gen_hook)

